compiler/packages/babel-plugin-react-compiler/src/Inference/InferEffectDependencies.ts

```diff
@@ -0,0 +1,60 @@
+import { ArrayExpression, Effect, Environment, FunctionExpression, GeneratedSource, HIRFunction, IdentifierId, Instruction, isUseEffectHookType, makeInstructionId } from "../HIR";
+import { createTemporaryPlace } from "../HIR/HIRBuilder";
+
+export function inferEffectDependencies(
+    env: Environment,
+    fn: HIRFunction,
+  ): void {
+    const fnExpressions = new Map<IdentifierId, FunctionExpression>();
+    for (const [, block] of fn.body.blocks) {
+      for (const instr of block.instructions) {
+        const {value, lvalue} = instr;
+        if (
+          value.kind === 'FunctionExpression'
+        ) {
+          fnExpressions.set(lvalue.identifier.id, value)
+        }    
+      }
+    }
+    
+    for (const [, block] of fn.body.blocks) {
+      let newInstructions = [...block.instructions];
+      let addedInstrs = 0;
+      for (const [idx, instr] of block.instructions.entries()) {
+        const {value} = instr;
+  
+        /*
+         * This check is not final. Right now we only look for useEffects without a dependency array.
+         * This is likely not how we will ship this feature, but it is good enough for us to make progress
+         * on the implementation and test it. 
+         */
+        if (
+          value.kind === 'CallExpression' &&
+          isUseEffectHookType(value.callee.identifier) &&
```

Comment:
might be worth handling other kinds too, eg `isUseLayoutEffectHookType` and `isUseInsertionEffectHookType`